### ISSUE TITLE
Installing OpenSTF on Cent Os 7 Minimal Raspberry Pi

### ISSUE THREAD

**Issue by caleb7773** [openstf/stf#820]
Has anyone successfully installed OpenSTF on a Raspberry Pi? And more specifically on Cent OS 7? I am not very Linux savy, but have done a few minor installs in the past and am trying to put OpenSTF on a Pi. Does anyone have instructions for how to go about doing that?
Thanks!

**Comment by thinkhy** [comment]
Can you install docker on Raspberry Pi?

**Comment by jachstet-sea** [comment]
@thinkhy : Yep, you can https://www.raspberrypi.org/blog/docker-comes-to-raspberry-pi/
However, the question is why not using the "local deployment" way only? Install node, "npm -g" install stf and run it?

**Comment by thinkhy** [comment]
Short answer is docker can offer much easier deployment and you needn't care about details of OS env.

**Comment by sorccu** [comment]
I would recommend the docker image which you can find on our docker hub page. STF and its dependencies have many native dependencies which can be a pain to get working. At the very least take a look at the Dockerfile in this repo’s docker folder for tips. I personally cannot help more than that so if you decide to go the local installation route it’ll be up to you to get it running. Feel free to post instructions for others here if you end up doing it, though.

**Comment by caleb7773** [comment]
I managed to get OpenSTF running on the Pi. Utilizing Pacman to get the dependencies and then NPM to install stf. I had to make some changes to the npm default folder because I was getting a lot of permissions errors. Now I am having trouble getting the phones to actually connect. What is required to be on the phones besides USB debugging enabled? If I can get it working I'll rebuild it with instructions.
Do I need to so something with the keys to enable the phones to connect? Thanks,


![image](https://user-images.githubusercontent.com/37095625/37260595-9a7db346-256b-11e8-9d7b-b6fcdcddcb17.png)

![image](https://user-images.githubusercontent.com/37095625/37260617-af61cb9e-256b-11e8-8a35-82f190a407a0.png)

**Comment by caleb7773** [comment]
Also I am wondering how people are running the RethinkDB and STF on the same machine, everytime I run RethinkDB I am unable to then run the STF Local command. My current work (due to being in a minimal OS) is to run RethinkDB on the actual Raspberry Pi terminal, then I SSH into the Raspberry Pi from another computer and run stf local --public-ip 192.168.60.23 --allow-remote
Thanks again,

**Comment by caleb7773** [comment]
Installing Arch Linux on a Raspberry Pi:
https://archlinuxarm.org/platforms/armv7/broadcom/raspberry-pi-2#qt-platform_tabs-ui-tabs2


# System Upgrade
pacman -Syu
# STF deps:
pacman -S android-tools nodejs npm graphicsmagick zeromq pkg-config 
# RethinkDB build dependencies
pacman -S base-devel # Select ALL
pacman -S libsodium libpgm protobuf protobuf-c libboost boost boost-libs python2 gperftools python2-pip
#### libboost was not found... so I installed "boost-libs"###
# Download rethinkDB sources
##################################################
I may have had to install curl with pacman but don't remember
#################################################

curl -O http://download.rethinkdb.com/dist/rethinkdb-2.0.4.tgz
tar zxvf rethinkdb-2.0.4.tgz
 # Change default python version 
ln -sf /usr/bin/python2.7 /usr/bin/python
# Configure and build! (it may take hours on Pi)
cd rethinkdb-2.0.4
./configure --with-system-malloc --allow-fetch
make

make install
# finally install stf

#################################################
Permissions issues found here:
https://docs.npmjs.com/getting-started/fixing-npm-permissions
#################################################

 mkdir ~/.npm-global
npm config set prefix '~/.npm-global'
 export PATH=~/.npm-global/bin:$PATH
 source ~/.profile ####THIS FAILED FOR ME#####
 npm install -g jshint
    NPM_CONFIG_PREFIX=~/.npm-global


pacman -S android-udev

npm install -g stf

To then run STF I had to run:

RethinkDB

Then I SSH'd into the pi from a different machine and issued this:

stf local --public-ip 192.168.60.23 --allow-remote

Then I was able to access it remotely at:
http://x.x.x.x:7100/

This is more or less what I did.

**Comment by caleb7773** [comment]
Also had to allow it through the firewall because it was being blocked.

**Comment by jachstet-sea** [comment]
You mean by running rethinkdb and stf on the same machine that once you start rethinkdb it blocks your terminal and you cannot execute any more commands on that terminal unless you kill rethinkdb?
That would also imply that if you log out of the Pi or lose the network connection, STF gets terminated?
If so, have a look at "screen": https://www.howtoforge.com/linux_screen. It lets commands continue to run and be sent to the background so you can keep on working on the same terminal and/or log off

**Comment by thinkhy** [comment]
Or you can use the command of "nohup rethinkdb &"

**Comment by gounthar** [comment]
Made it with docker and docker-compose on a Raspberry 3B with latest Raspbian.
```
pi@openstf-provider:~ $ uname -a
Linux openstf-provider 4.9.80-v7+ #1098 SMP Fri Mar 9 19:11:42 GMT 2018 armv7l GNU/Linux
```
```
pi@openstf-provider:~ $ docker-compose version
docker-compose version 1.13.0dev, build ae2cc6b
docker-py version: 2.2.1
CPython version: 2.7.13
OpenSSL version: OpenSSL 1.0.1t  3 May 2016
```
```
pi@openstf-provider:~ $ docker version
Client:
 Version:       18.03.0-ce
 API version:   1.37
 Go version:    go1.9.4
 Git commit:    0520e24
 Built: Wed Mar 21 23:18:35 2018
 OS/Arch:       linux/arm
 Experimental:  false
 Orchestrator:  swarm

Server:
 Engine:
  Version:      18.03.0-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.4
  Git commit:   0520e24
  Built:        Wed Mar 21 23:14:37 2018
  OS/Arch:      linux/arm
  Experimental: false

```
For the time being, it works flawlessly (except the limitations due to the device).
![screenshot-2018-3-23 stf](https://user-images.githubusercontent.com/116569/37855560-22105074-2ef0-11e8-87b1-866917572c44.png)

**Comment by catalin-florescu** [comment]
@gounthar can you provide the steps you made to install it, please?
